Radix history unit tests

maximus19910112 opened the pull request:
## Overview
Creating a test specification for RadixTree implementation

### Notes
For writing tests, I used examples from module [HistoryActionTests.scala](https://github.com/maximus19910112/rchain/blob/c2535eb8ce3d65cc417a1e56af9ae70e1af647c9/rspace/src/test/scala/coop/rchain/rspace/history/HistoryActionTests.scala).
I appended unit-tests, that validate work of main procedures of module [RadixTree.scala](https://github.com/maximus19910112/rchain/blob/c2535eb8ce3d65cc417a1e56af9ae70e1af647c9/rspace/src/main/scala/coop/rchain/rspace/history/RadixTree.scala):
1) RadixTreeImpl.update:
  * Creating tree with one leaf
  * Creating node with two leafs
  * Creating tree consisting of 1 nodePtr and 2 leafs
  * Updating leaf data
  * Assertion check: update does not allow to enter keys with different lengths
  * Assertion check: update does not allow to radix key is smaller than NodePtr key
  * Check: deleting non - existent data should return none


2) RadixTreeImpl.delete:
  * Deleting non-existent data returns none
  * Destroy tree that consists of 1 node should destroy tree
  * Delete leaf from node with 2 leafs leaves node with 1 leaf
  * Delete leaf from tree with 1 nodePtr and 2 leafs leaves tree with 1 leaf(NodePtr is also destroyed)

3) RadixTreeImpl.read:
  * Reading existing data should return data(in case data don't exist it should return none)
  
4) RadixTreeImpl.makeActions:
  * Function should build tree with many leafs and nodePtr's correctly
  * Assertion check: function should detect collision in KVDB
  * Function should not create artefacts (no creating excess nodes while it works)

5) RadixTreeImpl.saveNode: 
  * Function should save selected node into store

6) Functions RadixTree.Codecs.encode and RadixTree.Codecs.decode should work correctly
7) Function HistoryMergingInstances.c


### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

tgrospic committed:
Remove isReceived method from BlockRetriever trait

tgrospic committed:
Move old LMDBStore implementation to rspace-bench project

stanislavlyalin commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeTests.scala:
```diff
@@ -0,0 +1,695 @@
+import scala.concurrent.duration._
+import scala.util.Random
+
+trait RadixTreeObject[F[_]] {}
```
This trait is not used so can be removed. You can just have an object, not a companion object (trait is not necessary).

stanislavlyalin commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeTests.scala:
```diff
@@ -0,0 +1,695 @@
+    } yield impl //RadixTreeObject(store)
+  def create[F[_]: Concurrent: Sync: Parallel](
+      store: KeyValueTypedStore[F, ByteVector, ByteVector]
+  ): F[RadixTreeImpl[F]] = RadixTreeObject(store)
```
I guess you can use just one of those functions.

stanislavlyalin commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeTests.scala:
```diff
@@ -0,0 +1,695 @@
+  def generateDataWithLastNonZeroByte(lastByte: Byte): Array[Byte] =
+    (List.fill(31)(0) ++ List.fill(1)(lastByte.toInt)).map(_.toByte).toArray
+
+  "tree with makeActions" should "be built correctly!!!" in withRadixTreeImplAndInMemoStore {
```
* I checked other tests in the project and in some cases sentence starts with a lowercase letter and in other cases with uppercase letters. I prefer to use uppercase.
* Also, you can change `should "be...` with `shouldBe "...`.
* I guess the `!!!` is not add additional information so can be removed.

stanislavlyalin commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeTests.scala:
```diff
@@ -0,0 +1,695 @@
+  def generateDataWithLastNonZeroByte(lastByte: Byte): Array[Byte] =
+    (List.fill(31)(0) ++ List.fill(1)(lastByte.toInt)).map(_.toByte).toArray
+
+  "tree with makeActions" should "be built correctly!!!" in withRadixTreeImplAndInMemoStore {
```
What is `tree with makeActions`? You can ignore my comment. Maybe I am just not familiar with logic.

maximus19910112 replied:
I wanted to clarify that in this test for creating tree used function makeActions. It is also possible to do this with the help of function RadixTreeImpl.update

stanislavlyalin commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeTests.scala:
```diff
@@ -0,0 +1,695 @@
+        dataForHashes = lastByteForHashes.map(byte => generateDataWithLastNonZeroByte(byte))
+        hashesBlake   = dataForHashes.map(hash => Blake2b256Hash.fromByteArray(hash))
+
+        insertActions2 = InsertAction(keys(0), hashesBlake(0)) ::
```
What index `2` in `insertActions2` means?

stanislavlyalin commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeTests.scala:
```diff
@@ -0,0 +1,695 @@
+          InsertAction(keys(3), hashesBlake(3)) :: Nil
+
+        newRootNodeOpt <- impl.makeActions(rootNode, insertActions2)
+        treeInfo       <- impl.printTree(newRootNodeOpt.get, "TREE1", false)
```
Maybe the name `printTree` should be changed to `toString` or `visualize` or `printToString` or something... I was just confused where the result is placed: in stdout or in string. This is just a nitpick, be free to ignore it.

stanislavlyalin commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeTests.scala:
```diff
@@ -0,0 +1,695 @@
+          "      [12]LEAF: prefix = empty, data = 0000...0002"
+        )
+
+        _ = treeInfo shouldBe etalonTree
```
Consider returning String from `printTree` so `etalonTree` can be just a String, not Vector.

stanislavlyalin commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeTests.scala:
```diff
@@ -0,0 +1,695 @@
+      } yield ()
+  }
+
+  "appending leaf in empty tree" should "work" in withRadixTreeImplAndInMemoStore {
```
What `should work` mean? Should compile or should work without exceptions or should produce the correct result? Consider changing to a more informative string.

stanislavlyalin commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeTests.scala:
```diff
@@ -0,0 +1,695 @@
+        printedTreeStr <- impl.printTree(node, "NODE AFTER SERIALIZE", false)
+
+        etalonString = "ByteVector(37 bytes, 0xff03f8aff100000000000000000000000000000000000000000000000000000000000000ad)"
+        printed      = println(s"Serialized vector : ${serializedNode.toString()}")
```
What is the purpose of this print in the test?

stanislavlyalin reviewed (commented):
Nice work! Some tips here.

You can use Markdown links to the files. It is more clear. For example, `[HistoryActionTests.scala](https://github.com/maximus19910112/rchain/blob/c2535eb8ce3d65cc417a1e56af9ae70e1af647c9/rspace/src/test/scala/coop/rchain/rspace/history/HistoryActionTests.scala)`. Here is result of this MD code: [HistoryActionTests.scala](https://github.com/maximus19910112/rchain/blob/c2535eb8ce3d65cc417a1e56af9ae70e1af647c9/rspace/src/test/scala/coop/rchain/rspace/history/HistoryActionTests.scala).

You can leave the PR template unchanged. I mean line `Please make sure that this PR:` and below.

You should fill some of the PRs fields, such as Assignees, Labels, Projects.

If PR is ready, you should mark it as ready-for-review.

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeTests.scala:
```diff
@@ -0,0 +1,695 @@
+  ): F[RadixTreeImpl[F]] = RadixTreeObject(store)
+}
+
+class RadixTreeTests extends FlatSpec with Matchers with OptionValues with InMemoryHistoryTestBase {
```
Please rename your file and class/object as RadixTreeSpec, because in other tests we use this style.

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeTests.scala:
```diff
@@ -0,0 +1,695 @@
+    (List.fill(31)(0) ++ List.fill(1)(lastByte.toInt)).map(_.toByte).toArray
+
+  "tree with makeActions" should "be built correctly!!!" in withRadixTreeImplAndInMemoStore {
+    (radixTreeImplF, typedStore) =>
```
typedStore not used in this test, please replace this to `(radixTreeImplF, _)`

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeTests.scala:
```diff
@@ -0,0 +1,695 @@
+    val radixTreeImpl =
+      RadixTreeObject.create(typedStore)
+    f(radixTreeImpl, store).runSyncUnsafe(20.seconds)
+  }
```
Let's delete RadixTreeObject
```scala
    val radixTreeImpl = new RadixTreeImpl[Task](typedStore)
    f(radixTreeImpl, store).runSyncUnsafe(20.seconds)
```

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeTests.scala:
```diff
@@ -0,0 +1,695 @@
+
+  protected def withRadixTreeImplAndInMemoStore(
+      f: (
+          Task[RadixTreeImpl[Task]],
```
In this case we don't need first Task,
```scala
f: (
          RadixTreeImpl[Task],
          InMemoryKeyValueStore[Task]
      ) => Task[Unit]
```
This is used in RadixHistory class because in function `apple` we call
<https://github.com/maximus19910112/rchain/blob/6fe380a90a804599dcf0a7360f81755dc7a31481/rspace/src/main/scala/coop/rchain/rspace/history/instances/RadixHistory.scala#L25>
but you don't do this.
Now we can delete these strings for all tests:
```scala
        impl        <- radixTreeImplF
```

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeTests.scala:
```diff
@@ -0,0 +1,695 @@
+  }
+
+  def createDeleteActions(keys: List[String]): List[DeleteAction] =
+    keys.map(key => DeleteAction(TestData.hexKey(key))) ++ Nil
```
What purpose of this `++ Nil`? `List.map` return List() without this operation. Perhaps you are confused with `val data = insert(_zeros) :: Nil`

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeTests.scala:
```diff
@@ -0,0 +1,695 @@
+  def createDeleteActions(keys: List[String]): List[DeleteAction] =
+    keys.map(key => DeleteAction(TestData.hexKey(key))) ++ Nil
+
+  protected def withRadixTreeImplAndInMemoStore(
```
Maybe better to use short naming for example: `withImplAndStore`?

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeTests.scala:
```diff
@@ -0,0 +1,695 @@
+    val res3     = HistoryMergingInstances.commonPrefix(v12345, emptyVal)
+    val res4     = HistoryMergingInstances.commonPrefix(v12345, v12367)
+    val res5     = HistoryMergingInstances.commonPrefix(v22345, v12345)
+    val res6     = HistoryMergingInstances.commonPrefix(emptyVal, emptyVal)
```
Please create test for RadixTree.commonPrefix function. We don't work with MergingHistory.

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeTests.scala:
```diff
@@ -0,0 +1,695 @@
+class RadixTreeTests extends FlatSpec with Matchers with OptionValues with InMemoryHistoryTestBase {
+  def generateDataWithLastNonZeroByte(lastByte: Byte): Array[Byte] =
+    (List.fill(31)(0) ++ List.fill(1)(lastByte.toInt)).map(_.toByte).toArray
+
```
Please move this function to the bottom of the page so that it is grouped with the place where it is used. This improves the readability of the code.
Please rewrite this function similar https://github.com/maximus19910112/rchain/blob/6fe380a90a804599dcf0a7360f81755dc7a31481/rspace/src/test/scala/coop/rchain/rspace/history/HistoryActionTests.scala#L276-L280 
you can use this name ```def createBV32(s: String): ByteVector```
You can use this code as example: https://github.com/rchain/rchain/blob/6fe380a90a804599dcf0a7360f81755dc7a31481/node/src/test/scala/coop/rchain/node/perf/HistoryGenKeySpec.scala#L194-L197

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeTests.scala:
```diff
@@ -0,0 +1,695 @@
+      keyAndData =>
+        (
+          TestData.hexKey(keyAndData._1),
+          Blake2b256Hash.fromByteArray(generateDataWithLastNonZeroByte(keyAndData._2))
```
`Blake2b256Hash.fromByteVector(createBV32(keyAndData._2))`

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeTests.scala:
```diff
@@ -0,0 +1,695 @@
+
+    convertedKeysAndDatas.map(insData => InsertAction(insData._1, insData._2)) ++ Nil
+  }
+
```
```scala
  def createInsertActions(
      tuplesKeyAndHash: List[(String, Byte)]
  ): List[InsertAction] =
    tuplesKeyAndHash.map { keyAndData =>
      val key  = hexKey(keyAndData._1)
      val data = Blake2b256Hash.fromByteArray(generateDataWithLastNonZeroByte(keyAndData._2))
      InsertAction(key, data)
    }
```

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeTests.scala:
```diff
@@ -0,0 +1,695 @@
+    println(s"PREFIX (1, 2, 3, 4, 5)(1, 2, 3, 6, 7) should be: $res4")
+    println(s"PREFIX (2, 2, 3, 4, 5)(1, 2, 3, 4, 5) should be: $res5")
+    println(s"PREFIX (empty)(empty) should be: $res6")
+    res1 shouldBe List(1, 2)
```
Please remove all print functions because we should don't create messages in consol during tests.

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeTests.scala:
```diff
@@ -0,0 +1,695 @@
+        commonKeyPartForTwoLeafs = "001122"
+        keys = Vector[Seq[Byte]](
+          TestData.hexKey(commonKeyPartForTwoLeafs + "013"),
+          TestData.hexKey(commonKeyPartForTwoLeafs + "225")
```
Please replace this and all same place to `createBV32(commonKeyPartForTwoLeafs + "013")))`

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeTests.scala:
```diff
@@ -0,0 +1,695 @@
+  }
+
+  def createInsertActions(
+      tuplesKeyAndHash: List[(String, Byte)]
```
`tuplesKeyAndHash: List[(String, String)]`

maximus19910112 replied:
Why?

hilltracer replied:
In this form, this function will be more convenient to use. But this requires changing its implementation.

tgrospic commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,726 @@
+  def createInsertActions(
+      tuplesKeyAndHash: List[(String, Byte)]
+  ): List[InsertAction] =
+    tuplesKeyAndHash.map { keyAndData =>
```
Prefer destructure tuple so each element can be named.
```scala
tuplesKeyAndHash.map { case (key, data) =>
```

tgrospic commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,726 @@
+        ByteVector(0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+          0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+          0x00, 0x00, 0x00, 0x33)
+      )
```
Instead of constructing binary data from separated bytes, more compact representation is a hex string.

tgrospic commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,726 @@
+    } yield ()
+  }
+
+  "updating leaf" should "work correctly" in withImplAndStore { (impl, _) =>
```
If possible. test description should say exactly what is doing. "Works correctly" should be the last resort.

tgrospic commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,726 @@
+    (impl, _) =>
+      val leafData    = createBV32(0xCB.toByte)
+      val leafKey     = TestData.hexKey("0123456F1").toVector
+      val testLeafKey = TestData.hexKey("112").toVector
```
@hilltracer Why is not possible to insert keys with different length if they are in different sub-root?

hilltracer replied:
It's possible. We have wrong Assertion error message. @maximus19910112  please rewrite this message to `RadixKey must not be a complete part of another radixKey in this tree.` In your unit tests and in Radix - implementation: https://github.com/hilltracer/rchain/blob/3991334c48ab79aa8b515896f26b74f6d85d12fc/rspace/src/main/scala/coop/rchain/rspace/history/RadixTree.scala#L757
Also please rewrite this message in a similar way: https://github.com/hilltracer/rchain/blob/3991334c48ab79aa8b515896f26b74f6d85d12fc/rspace/src/main/scala/coop/rchain/rspace/history/RadixTree.scala#L890

tgrospic reviewed (commented):
Good work @maximus19910112!

stanislavlyalin committed:
Merge branch 'dev' into feature/radix-history

stanislavlyalin committed:
Fixed compiler errors after merging latest `dev` into `feature/radix-history`

bors[bot] committed:
Merge #3620

3620: Sync dev with feature/radix-history r=stanislavlyalin a=stanislavlyalin



Co-authored-by: bors[bot] <26634292+bors[bot]@users.noreply.github.com>
Co-authored-by: hilltracer <hilltracer@yandex.ru>
Co-authored-by: Tomislav Grospic <grospic@gmail.com>
Co-authored-by: Will Qiu <zsluedem06@gmail.com>

tgrospic committed:
Update Scala 2.12.15, kind projector

tgrospic committed:
Replace type param ? with * (Scala 3 compat)

tgrospic committed:
build.sbt - move kamon override to common overrides

tgrospic committed:
Update dependency overrides, cats-tagless-macros, scala-logging

tgrospic committed:
Update monix 3.4.0

tgrospic committed:
Add dependency to scala-collection-compat (instead of override)

tgrospic committed:
Update sbt-scapegoat plugin 1.1.1 (1.4.11)

tgrospic committed:
Merge branch 'feature/hard-fork-2' into dev

# Conflicts:
#	casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala
#	casper/src/test/scala/coop/rchain/casper/util/rholang/RuntimeSpec.scala
#	rspace/src/main/scala/coop/rchain/rspace/state/RSpaceImporter.scala
#	rspace/src/test/scala/coop/rchain/rspace/history/HistoryRepositoryGenerativeSpec.scala

tgrospic committed:
Fix exporter validation tests (channels map removed)

bors[bot] committed:
Merge #3616 #3619

3616: Fix sorting match case (freeCount field) r=tgrospic a=tgrospic



3619: Merge Rholang 1.1 to dev branch r=tgrospic a=stanislavlyalin



Co-authored-by: Tomislav Grospic <grospic@gmail.com>
Co-authored-by: Joseph Denman <josephrileydenman@gmail.com>
Co-authored-by: bors[bot] <26634292+bors[bot]@users.noreply.github.com>
Co-authored-by: Will Qiu <zsluedem06@gmail.com>
Co-authored-by: will <zsluedem06@gmail.com>
Co-authored-by: stanislavlyalin <lyalinstas@gmail.com>

tgrospic committed:
Merge branch 'dev' into dev-hf2

tgrospic committed:
Fix consume binding in PoSTest

tgrospic committed:
Fix hard-coded state hashes (after Rholang 1.1 changes)

tgrospic committed:
Fix kryo tests (deserialized empty List mapping error)

bors[bot] committed:
Merge #3622

3622: Merge of feature/hard-fork-2 to dev branch r=tgrospic a=tgrospic



Co-authored-by: nutzipper <1746367+nzpr@users.noreply.github.com>
Co-authored-by: nzpr <1746367+nzpr@users.noreply.github.com>
Co-authored-by: Tomislav Grospic <grospic@gmail.com>
Co-authored-by: bors[bot] <26634292+bors[bot]@users.noreply.github.com>
Co-authored-by: Will Qiu <zsluedem06@gmail.com>
Co-authored-by: will <zsluedem06@gmail.com>

bors[bot] committed:
Merge #3624

3624: Transaction balance script fix r=zsluedem a=zsluedem



Co-authored-by: Will Qiu <zsluedem06@gmail.com>

zsluedem committed:
consider genesis vaul include CoopMultiSigVault

bors[bot] committed:
Merge #3621

3621: Update Scala to version 2.12.15 r=tgrospic a=tgrospic



Co-authored-by: Tomislav Grospic <grospic@gmail.com>

zsluedem committed:
add a testcase to reproduce error in https://github.com/rchain/rchain/pull/3418#issuecomment-1066275032

zsluedem committed:
fix treeHashMap init leaf node bug

stanislavlyalin committed:
Added shardId to DeployDataProto

tgrospic committed:
Add sdk project to build.sbt

tgrospic committed:
Initial DagManager, DagView and DagData interface definitions

tgrospic committed:
Add initial extensions for DagView and DagData

tgrospic committed:
Add initial Casper extensions for DagView and DagData

tgrospic committed:
Add initial BlockRequester interface

tgrospic committed:
Add initial implementation for NetworkBlockRequester

tgrospic committed:
Add initial implementation for RNodeDagManager

tgrospic committed:
Add initial function to create RNode state (using DagManager)

stanislavlyalin committed:
Updated related code after adding `shardId` to DeployData

zsluedem committed:
uncomment tests and add comments

stanislavlyalin committed:
TestNode read shardId from supplied `genesis` context

stanislavlyalin committed:
Eliminated passing the `shardId` to `BlockCreator`

stanislavlyalin committed:
For exploratory deploys `shardId` can be empty. Consequently removed unnecessary passing `shardId` and from RuntimeManager also

stanislavlyalin committed:
Fixed DeployData serialization

bors[bot] committed:
Merge #3626

3626: Transaction balance fix r=zsluedem a=zsluedem



Co-authored-by: Will Qiu <zsluedem06@gmail.com>

zsluedem committed:
remove used code tryIt in TreeHashMapTest.rho

stanislavlyalin committed:
Fixed test `Secp256k1 should be valid signature algorithm to sign a deploy`: added handling of `shardId`

stanislavlyalin committed:
Fixed test `StoragePrinter should print unmatched sends of multiple deploys`: after adding `shardId` field deploys signatures were updated

stanislavlyalin committed:
Arbitrary string is used for `shardId`

stanislavlyalin committed:
`shardId` is set as an empty string for MergeabilityRules because it is not important in this case

stanislavlyalin committed:
Added shard check when creating a deploy

tgrospic committed:
Fix result type for loading method for sender

tgrospic committed:
Combine finalization info together when message is inserted

tgrospic committed:
Removed stake field from DagData, added mid, and sid accessors

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,726 @@
+        ("FF0000201", 0xB.toByte),
+        ("FF002111", 0x1.toByte),
+        ("FF002112", 0x2.toByte)
+      )
```
Please use dataSet for this test as show in RadixSpecification: https://github.com/rchain/rchain/issues/3632.
![image](https://user-images.githubusercontent.com/81892279/158965451-d8ccf142-287b-4b5e-a021-d0110bf2f9d5.png)

tgrospic committed:
Add sdk project to root rchain project in build.sbt

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,726 @@
+import scala.util.Random
+
+class RadixTreeSpec extends FlatSpec with Matchers with OptionValues with InMemoryHistoryTestBase {
+  "tree with makeActions" should "be built correctly" in withImplAndStore { (impl, _) =>
```
Please group all makeAction function tests in one place. Preferably somewhere at the end, before export testing.

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,726 @@
+package coop.rchain.rspace.history
```
Please repare all warnings, weak warnings and typos.
![image](https://user-images.githubusercontent.com/81892279/158966661-fa96f467-6151-41bd-a0d2-a91a22e07d96.png)
You can find it in this area:
![image](https://user-images.githubusercontent.com/81892279/158966901-4e572acd-a1f6-4fcd-8331-006cdd8616dd.png).

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,726 @@
+
+import java.nio.ByteBuffer
+import scala.concurrent.duration._
+import scala.util.Random
```
Use `Optimize Imports` function in idea: Code -> Optimize Imports

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,679 @@
+      val keys = Vector[Seq[Byte]](
+        TestData.hexKey(commonKeyPartForTwoLeafs + "013"),
+        TestData.hexKey(commonKeyPartForTwoLeafs + "225")
+      )
```
Please use a prefix that is a multiple of 1 byte. (For this Hex code should be a multiple of 2 symbols: 0x0013, 0x0225)

stanislavlyalin committed:
Added shard check for all block's deploys when replaying a block

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,679 @@
+    res6 shouldBe (ByteVector.empty, ByteVector.empty, ByteVector.empty)
+  }
+
+  def createBV32(lastByte: Byte): ByteVector =
```
Let's change input type for `lastByte` to Int. In this case you can change `createBV32(0xXX.toByte)`->`createBV32(0xXX)` in many places.

hilltracer replied:
better use String instead of Int. In this case you can create function similar: 
```scala
def createBV(s: String): ByteVector = ByteVector(Base16.unsafeDecode(s))
```
But with fills 32-n bytes as 0x00.

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,679 @@
+      for {
+        item1 <- impl.update(
+                  RadixTree.EmptyItem,
+                  TestData.hexKey("FFFFFFF1").toVector,
```
Instead of this construction let's create new function 
```scala
def createBV(s: String): ByteVector = ByteVector(Base16.unsafeDecode(s))
```
Please use this function in all places.

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,679 @@
+  }
+
+  def createBV32(lastByte: Byte): ByteVector =
+    (List.fill(31)(0) ++ List.fill(1)(lastByte.toInt)).map(_.toByte).toVector
```
:-)
```scala
  def createBV32(lastByte: Int): ByteVector =
    ByteVector((List.fill(31)(0) :+ lastByte).map(_.toByte))
```

hilltracer replied:
```scala
 def createBV32(lastByte: String): ByteVector =
{
    nonEmpty = createBV(s)
    empty = ByteVector (List.fill(32-nonEmpty.size)(0))
    empty ++ nonEmpty
}
```

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,679 @@
+
+  "appending leaf to tree with one leaf " should "create 2 leafs with node ptr" in withImplAndStore {
+    (impl, _) =>
+      val commonKeyPartForTwoLeafs = "001122"
```
What's porpose of `commonKeyPartForTwoLeafs`? Let's add this value in key directly 0011220013, 0011220225.

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,679 @@
+      } yield ()
+  }
+
+  "appending leaf to tree with one leaf " should "create 2 leafs with node ptr" in withImplAndStore {
```
`create 2 leafs with node ptr` - this phrase correct only if leafs contain the same part of prefix.

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,679 @@
+        _ = printedTree2 shouldBe etalonTree2
+      } yield ()
+  }
+  "appending leaf to leaf" should "create node with two leafs" in withImplAndStore { (impl, _) =>
```
this phrase correct only if leafs don't contain the same part of prefix.

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,679 @@
+      }
+  }
+
+  "RadixTreeImpl" should "not allow to radix key is smaller than NodePtr key" in withImplAndStore {
```
not allow to paste radix key smaller than..

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,679 @@
+        ex shouldBe a[AssertionError]
+        ex.getMessage shouldBe s"assertion failed: Radix key should be longer than NodePtr key."
+      }
+  }
```
that's better
```scala
    (impl, _) =>
      for {
        err <- impl
                .update(
                  NodePtr(ByteVector(0x00, 0x11), createBV32(0x00.toByte)),
                  TestData.hexKey("0"),
                  createBV32(0x33.toByte)
                )
                .attempt
      } yield {
        err.isLeft shouldBe true
        val ex = err.left.get
        ex shouldBe a[AssertionError]
        ex.getMessage shouldBe s"assertion failed: Radix key should be longer than NodePtr key."
      }
```

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,679 @@
+      printedTreeStr <- impl.printTree(rootNode, "TREE (TEST DELETE NOT EXISTING LEAF)", true)
+
+      //  Trying to delete not existing leaf...
+      del <- impl.delete(itemOpt.get, TestData.hexKey("000").toVector.tail)
```
Please use multiples of bytes values "0000"

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,679 @@
+      itemOpt  <- impl.update(RadixTree.EmptyItem, leafKey, leafData)
+      rootNode <- impl.constructNodeFromItem(itemOpt.get)
+
+      printedTreeStr <- impl.printTree(rootNode, "TREE (TEST DELETE NOT EXISTING LEAF)", true)
```
What is porpose of printing tree? Let's delete this string.

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,679 @@
+        itemOpt   <- impl.update(RadixTree.EmptyItem, leafKey, leafData)
+        rootNode1 <- impl.constructNodeFromItem(itemOpt.get)
+
+        printedTreeStr <- impl.printTree(rootNode1, "TREE (TEST DELETING ONE LEAF)", true)
```
Please compare results of printing with etalon.

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,679 @@
+        rootNode2           = rootNode1.updated((leafKey.head).toInt, deletedItem.get)
+        printedEmptyTreeStr <- impl.printTree(rootNode2, "EMPTY TREE", true)
+
+        emptyTreeStr = Vector("EMPTY TREE: root =>")
```
Please change name of `emptyTreeStr` to `etalonEmptyTreeStr`.

stanislavlyalin committed:
Provided additional information about shard names in the error message in BlockAPI.deploy

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,679 @@
+      } yield ()
+  }
+
+  "deleting data from leaf" should "destroy this leaf" in withImplAndStore { (impl, _) =>
```
deleting one leaf from a child node containing two leafs should delete child node and reformat parrent node

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,679 @@
+
+      readedDataOpt <- impl.read(rootNode, key)
+
+      _ = readedDataOpt.map(readedData => readedData.toArray shouldBe itemData.toArray)
```
`      _ = readedDataOpt.get shouldBe itemData`

maximus19910112 replied:
Maybe it's better not to use .get, for safety?

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,679 @@
+      notExistingKey = TestData.hexKey("000").toVector
+      readedDataOpt  <- impl.read(rootNode, notExistingKey)
+
+      _ = readedDataOpt.map(readedData => readedData shouldBe none)
```
This check has never been performed because readedDataOpt is None
use `_ = readedDataOpt shouldBe none`

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,679 @@
+
+      newRootNodeOpt2 <- impl.makeActions(newRootNodeOpt1.get, deleteRecord)
+      hashOpt         = newRootNodeOpt2.map(node => impl.saveNode(node))
+      err             <- impl.commit.attempt
```
Such better
```scala
    val collisionKVPair = (copyBVToBuf(History.emptyRootHash.bytes), ByteVector(0x00))
    for {
      _ <- inMemoStore.put[ByteVector](Seq(collisionKVPair), copyBVToBuf)
      _   = impl.saveNode(emptyNode)
      err <- impl.commit.attempt
```

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,679 @@
+      ex shouldBe a[RuntimeException]
+      ex.getMessage shouldBe
+        s"1 collisions in KVDB (first collision with key = " +
+          s"${History.emptyRootHash.bytes.toHex})."
```
please move this checks to `for` body.

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,679 @@
+        s"1 collisions in KVDB (first collision with key = " +
+          s"${History.emptyRootHash.bytes.toHex})."
+
+      println(ex.getMessage)
```
Please delete

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,679 @@
+  }
+
+  "encode and decode" should "give initial node" in withImplAndStore { (impl, _) =>
+    for {
```
please use data from example of RadixTree specification:
![image](https://user-images.githubusercontent.com/81892279/158988901-2628e063-a69a-4114-989c-c6d37e4555bf.png)

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,679 @@
+      etalonString = "ByteVector(37 bytes, 0xff03f8aff100000000000000000000000000000000000000000000000000000000000000ad)"
+
+      _ = deserializedNode shouldBe node
+      _ = serializedNode.toString() shouldBe etalonString
```
Please use validate without string in this test
```scala
      etalonData = createBV("ff03f8aff100000000000000000000000000000000000000000000000000000000000000ad")

      _ = deserializedNode shouldBe node
      _ = serializedNode shouldBe etalonData
```

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,679 @@
+              )
+
+      node           <- impl.constructNodeFromItem(item1.get)
+      printedTreeStr <- impl.printTree(node, "NODE BEFORE DECODING", true)
```
please delete printing

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,679 @@
+
+      deserializedNode = RadixTree.Codecs.decode(serializedNode)
+
+      printedTreeStr <- impl.printTree(node, "NODE AFTER SERIALIZE", true)
```
please delete printing

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,679 @@
+    } yield ()
+
+  }
+
```
Please append test with decode wrong serialized data. Also please change existing RadixTree implementation in this place: https://github.com/hilltracer/rchain/blob/3991334c48ab79aa8b515896f26b74f6d85d12fc/rspace/src/main/scala/coop/rchain/rspace/history/RadixTree.scala#L204
instead of 
```scala
decodeItem(0, emptyNode)
```
paste
```scala
try {
        decodeItem(0, emptyNode)
      } catch {
        case _: Exception =>
          assert(assertion = false, "Error during deserialization: invalid data format")
          emptyNode
      }
```
@tgrospic  do you agree this temporally solution?

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,679 @@
+  def createBV32(lastByte: Byte): ByteVector =
+    (List.fill(31)(0) ++ List.fill(1)(lastByte.toInt)).map(_.toByte).toVector
+  def createInsertActions(
+      tuplesKeyAndHash: List[(String, Byte)]
```
such better `List[(String, String)]`

nzpr committed:
Re-rnable block retriever spec

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,679 @@
+
+  }
+
+  "function makeActions" should "not create artefacts" in withImplAndStore { (impl, inMemoStore) =>
```
such better: `should not create artefacts in KV-store`

hilltracer replied:
Instead of all these makeactions tests, let's do the 1 test, with this Actions
1 action - With the help of makeactions to build a tree with a set of data as in the example to the specification: 0x111122334455 0x00..0001, 0x11112233aabb 0x00..0002, 0x1111Aabbcc 0x00..0003, 0x33 0x00..0004, 0xFF0011 0x00..0005, 0xFF012222 0x00 ..0006.
2 action - Print this tree and check with etalon
3 action - Validated that the number of nodes in Store is equal to 4
4 action - Delete all data from this tree with help MakeAction()
5 action - print empty tree, compare with etalon
6 action - Validated that the number of nodes in Store is equal to 5 (4 - old data + 1 empty node).

nzpr committed:
Fix

stanislavlyalin committed:
Provided default empty value to `shardId` in ConstructDeploy. It allowed elimination passing this parameter in tests

stanislavlyalin committed:
Removed `hidden` parameter from CLI deploy privateKey option to be visible in help message

stanislavlyalin committed:
Provided default empty value for `shardId` CLI option to fix runtime error `java.util.NoSuchElementException: None.get`

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,679 @@
+      } yield ()
+  }
+
+  "function saveNode" should "put node into store" in withImplAndStore { (impl, inMemoStore) =>
+    val itemData = createBV32(0xCB.toByte)
+    val key      = TestData.hexKey("0123456F1").toVector
+    for {
+      itemOpt     <- impl.update(RadixTree.EmptyItem, key, itemData)
+      nodesCount1 = inMemoStore.numRecords()
+
+      node     <- impl.constructNodeFromItem(itemOpt.get)
+      saved    = impl.saveNode(node)
+      commited <- impl.commit
+
+      //  After saving node numRecords must return 1
+      nodesCount2 = inMemoStore.numRecords()
+      _           = nodesCount1 shouldBe 0
+      _           = nodesCount2 shouldBe 1
+    } yield ()
+  }
```
```scala
  "function saveNode" should "put node into store" in withImplAndStore { (impl, inMemoStore) =>
    for {
      nodesCount1 <- Sync[Task].delay(inMemoStore.numRecords())
      hash        = impl.saveNode(emptyNode)
      _           <- impl.commit

      //  After saving node numRecords must return 1
      nodesCount2 = inMemoStore.numRecords()
      _           = nodesCount1 shouldBe 0
      _           = nodesCount2 shouldBe 1
    } yield ()
  }

  "function loadNode" should "load node from store" in withImplAndStore { (impl, inMemoStore) =>
    for {
      hash        <- Sync[Task].delay(impl.saveNode(emptyNode))
      _           <- impl.commit

      _        = impl.clearReadCache()
      _        = impl.clearWriteCache()
      loadNode <- impl.loadNode(hash)
      _        = loadNode shouldBe emptyNode
    } yield ()
  }
```

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,679 @@
+      )
+    )
+
+    val exportSettings = ExportDataSettings(
+      flagNodePrefixes = false,
+      flagNodeKeys = true,
+      flagNodeValues = true,
+      flagLeafPrefixes = false,
+      flagLeafValues = true
```
Please set all flags, and validate that all exported data is correct. For validating LeafPrefixes and LeafData you can just compare this value with initial data

hilltracer replied:
To validate NodePrefixes, NodeKeys and NodeValues don't use hardcoded values. Use validation procedure as here: https://github.com/hilltracer/rchain/blob/709e7fa710df0ef40b4a5723a2125709b533283f/node/src/test/scala/coop/rchain/node/perf/HistoryGenKeySpec.scala#L280-L435

Make the first export, then create storage from the exported data, and repeat the export from the new storage. If the second export is successful - this will be proof of the correctness of the node data (without hardcoded value).

hilltracer reviewed (commented):
Good luck!

bors[bot] committed:
Merge #3630

3630: SDK project with initial Dag Manager abstraction r=tgrospic a=tgrospic



Co-authored-by: Tomislav Grospic <grospic@gmail.com>

stanislavlyalin committed:
Fixed unit tests

bors[bot] committed:
Merge #3629

3629: Fix rholang treeHashMap `get`, `set` on the same key at the same time bug r=zsluedem a=zsluedem



Co-authored-by: Will Qiu <zsluedem06@gmail.com>

stanislavlyalin committed:
Updated pyrchain from 0.3.10 to 0.3.13a0

stanislavlyalin committed:
Added ability to provide `--shard-name` CLI option to RNode with empty string as default

nzpr committed:
remove unnecessary flag

stanislavlyalin committed:
Integration tests: used shard_id = 'test' for test_dag_correctness

stanislavlyalin committed:
Integration tests: used shard_id = 'test' for test_finalization

stanislavlyalin committed:
Integration tests: used shard_id = 'test' for test_slash

stanislavlyalin committed:
Integration tests: used shard_id = 'test' for test_synchrony_constraint

stanislavlyalin committed:
Integration tests: fixed remained tests

stanislavlyalin committed:
Added configuration parameter for PoS public keys

stanislavlyalin committed:
Config parameter `public-keys` renamed to `pos-multi-sig-public-keys`

stanislavlyalin committed:
Added configuration parameter `pos-multi-sig-quorum`. Parameter `publicKeys` renamed to `posMultiSigPublicKeys` in according to `posMultiSigQuorum`

stanislavlyalin committed:
Passed `posMultiSigQuorum` to `ProofOfStake`

stanislavlyalin committed:
Updated PoS contract with parameters `posMultiSigPublicKeys` and `posMultiSigQuorum`

stanislavlyalin committed:
Values for `posMultiSigQuorum` are the same as was hard-coded in PoS contract

stanislavlyalin committed:
Throw exception if `pos-multi-sig-quorum` has incorrect value

tgrospic committed:
Fix multi-sig vault with public key authentication

stanislavlyalin committed:
`throw` changed to `System.exit` because `throw` is disabled

stanislavlyalin committed:
Default shard id moved to a global variable in rnode.py so setting this parameter in tests doesn't need

stanislavlyalin committed:
Fixed using macros in PoS contract

stanislavlyalin committed:
Added descriptions for new macro parameters in the top of PoS.rhox

stanislavlyalin committed:
Fixed integration tests

bors[bot] committed:
Merge #3639

3639: Fix multi-sig vault with public key authentication r=tgrospic a=tgrospic



Co-authored-by: Tomislav Grospic <grospic@gmail.com>

stanislavlyalin committed:
Reverted changes in blockimplicits.scala. Used arbitrary value for shardId instead.

tgrospic committed:
Rename public keys channel in verify method

tgrospic committed:
Fix to use public key in nonce confirmation instead of auth token

bors[bot] committed:
Merge #3631

3631: Add shard id to deploy data r=stanislavlyalin a=stanislavlyalin



Co-authored-by: stanislavlyalin <lyalinstas@gmail.com>

stanislavlyalin committed:
Merge branch 'dev' into create-pub-key-configuration

Conflicts:
	casper/src/main/scala/coop/rchain/casper/engine/BlockApproverProtocol.scala
	casper/src/test/scala/coop/rchain/casper/engine/BlockApproverProtocolTest.scala

bors[bot] committed:
Merge #3638

3638: Added configuration parameters for PoS contract r=stanislavlyalin a=stanislavlyalin



Co-authored-by: stanislavlyalin <lyalinstas@gmail.com>

zsluedem committed:
add toHexString in ByteArrayOps

zsluedem committed:
add phloLimit in sourceDeployNow

zsluedem committed:
add ShortAnd and ShortOr in bnfc file

zsluedem committed:
add EShortAnd and EShortOr in models RhoTypes.proto

zsluedem committed:
add EShortAnd and EShortOr in normalizer and reducer

zsluedem committed:
add according implicits for EShortAnd and EShort Or

zsluedem committed:
add EShortAnd and EShortOR in ExprSortMatcher.scala

zsluedem committed:
add short circuit boolean test cases

hilltracer committed:
add short boolean in prettyPrinter

hilltracer committed:
add short boolean in Substitute

bors[bot] committed:
Merge #3643

3643: Fix nonce confirmation to use public key r=tgrospic a=tgrospic



Co-authored-by: Tomislav Grospic <grospic@gmail.com>

zsluedem committed:
make test comments more descriptive for short circuit boolean test cases

zsluedem committed:
add tests for validating substitution of short circuit syntax &&,||

zsluedem committed:
remove unused import in ShortCircuitBooleanSpec.scala

zsluedem committed:
clean test case in ShortCircuitBooleanSpec.scala

bors[bot] committed:
Merge #3649

3649: Add short-circuit evaluation boolean syntax   r=zsluedem a=zsluedem



Co-authored-by: Will Qiu <zsluedem06@gmail.com>
Co-authored-by: hilltracer <hilltracer@yandex.ru>

zsluedem committed:
replace and with && syntax in MultiSigRevVault.rho

zsluedem committed:
make CompiledRholangTemplate added comment in the end

zsluedem committed:
add POS MultiSig vault transfer test case

zsluedem committed:
rename test case in MultiSigRevVaultTest.rho

zsluedem committed:
add pubkey test cases in MultiSigRevVaultTest.rho

stanislavlyalin committed:
Removed BlockStore and LastFinalized stuff

hilltracer committed:
Update tree description.

stanislavlyalin committed:
Simplified getApprovedBlock implementation

stanislavlyalin committed:
Simplified codec for ApprovedStore key

stanislavlyalin committed:
Objects BlockStore and ApprovedStore removed to blockStore and approvedStore

stanislavlyalin committed:
Reverted explicit BlockStore and ApprovedStore passing

stanislavlyalin committed:
Removed duplicated implicit CasperBufferStorage[F] in BlockProcessor

bors[bot] committed:
Merge #3648

3648: Fix MultiSigVault boolean bug and add test cases r=zsluedem a=zsluedem



Co-authored-by: Will Qiu <zsluedem06@gmail.com>

stanislavlyalin committed:
Added extension for `put(b: BlockMessage)`

stanislavlyalin committed:
Fixed encoding/decoding for BlockStore and ApprovedStore

nzpr committed:
remove LFS storage

nzpr committed:
remove finalizer

stanislavlyalin committed:
Added BlockAPI_v2 trait

stanislavlyalin committed:
Added BlockAPI_v2 implementation

nzpr committed:
Remove estimator and safety oracle

stanislavlyalin committed:
Eliminated dependency from MultiParentCasper

stanislavlyalin committed:
Eliminated dependency from EngineCell

stanislavlyalin committed:
Eliminated dependency from SafetyOracle

nzpr committed:
Make explore deploy test not depend on finalization

nzpr committed:
Igore finalizer test

nzpr committed:
Make bonded status test not relying on finalization

nzpr committed:
Ignore LFDAN tests

nzpr committed:
Ignore MultiParentCasperFinalizationSpec

nzpr committed:
Ignore SingleParentCasperSpec

nzpr committed:
Ignore GHOST integration test

zsluedem committed:
make pos authkey not global and initialize in every transfer

tgrospic committed:
Fix for syntax import from shared package and BooleanF

stanislavlyalin committed:
Added scodec for ApprovedBlock

tgrospic committed:
Remove extra space

tgrospic committed:
Remove ProtoUtil.getMainChainUntilDepth

stanislavlyalin committed:
Removed unused case class

tgrospic committed:
Expose height map in BlockDagRepresentation

tgrospic committed:
Fix for listen for data at name to work without main parent

tgrospic committed:
Enable listen for data at name tests

tgrospic committed:
Restore "estimate parent properly" because it passes

tgrospic committed:
Temporary ignore integration test for finalization (until new finalizer is included)

stanislavlyalin committed:
Fixed imports cats syntax

bors[bot] committed:
Merge #3655

3655: BlockStore implemented as KeyValueTypedStore r=stanislavlyalin a=stanislavlyalin



Co-authored-by: stanislavlyalin <lyalinstas@gmail.com>
Co-authored-by: Tomislav Grospic <grospic@gmail.com>

bors[bot] committed:
Merge #3663

3663: Fix pos authKey global reused problem r=tgrospic a=zsluedem



Co-authored-by: Will Qiu <zsluedem06@gmail.com>

tgrospic committed:
Remove duplicated reference to validator identity in TestNode

tgrospic committed:
Supply validator identity to BlockReportAPI

tgrospic committed:
Validator identity supplied directly to CasperLaunch

tgrospic committed:
Temporary ignore REMINDER integration tests (until new finalizer is included)

tgrospic committed:
Removed EquivocationDetector and EquivocationsTracker

tgrospic committed:
Removed MultiParentCasper.normalizedInitialFault

tgrospic committed:
Removed Casper instance from BlockAPI get (light)block functions

tgrospic committed:
Add BlockDagStorage dependency to API servers

stanislavlyalin committed:
ValidatorIdentity passed directly to BlockAPIImpl

stanislavlyalin committed:
Synchronized with latest dev

bors[bot] committed:
Merge #3661

3661: Remove deprecated code (estimator, safety oracle) r=tgrospic a=nzpr



Co-authored-by: nutzipper <1746367+nzpr@users.noreply.github.com>
Co-authored-by: Tomislav Grospic <grospic@gmail.com>

tgrospic committed:
Remove Casper instance from BlockReportAPI

tgrospic committed:
Remove source info in BlockStore.getUnsafe extension

tgrospic committed:
Remove Casper from blocks related functions on BlockAPI

tgrospic committed:
Remove Casper instance from BlockAPI.getDataAtPar

tgrospic committed:
Move BlockDagKeyValueStorage to Casper project

tgrospic committed:
Removed EquivocationTrackerStore and EquivocationRecord

tgrospic committed:
Simplify scodec for BlockDagStorage (remove size prefix)

tgrospic committed:
Reformat BlockAPI.toposortDag

tgrospic committed:
Remove EngineCell (casper) from BlockAPI listen at name

tgrospic committed:
Remove EngineCell from BlockReportAPI

nzpr committed:
remove duplicate test

tgrospic committed:
Optimize BlockStore extensions to prevent boxing

bors[bot] committed:
Merge #3634

3634: Fix and re-enable block retriever spec r=nzpr a=nzpr



Co-authored-by: nutzipper <1746367+nzpr@users.noreply.github.com>

tgrospic committed:
Restore depth limit check in BlockAPI LFDAN methods

bors[bot] committed:
Merge #3664

3664: Casper (equivocation) cleanup (BlockAPI/BlockReportAPI) r=tgrospic a=tgrospic



Co-authored-by: Tomislav Grospic <grospic@gmail.com>

stanislavlyalin committed:
Renamed class fields according to convention https://docs.scala-lang.org/style/naming-conventions.html#constants-values-and-variables

nzpr committed:
Remove estimator

maximus19910112 committed:
Refactoring of RadixTree and RadixHistory:
 1) Append to RadixTree module value emptyRootHash
 2) Changed function name: changingTree replaced with saveAndCommit
 3) Little refactoring of emptyRootHash
 4) Refactoring of Codecs.decode method

Original exception supplied to wrapped exception

maximus19910112 committed:
Creating unit - tests

bors[bot] committed:
Merge #3667

3667: Remove estimator r=nzpr a=nzpr



Co-authored-by: nutzipper <1746367+nzpr@users.noreply.github.com>

nzpr committed:
Replace BlockDagRepresentation[F[_]] with pure DagRepresentation

nzpr committed:
Use DagRepresentation

nzpr committed:
adress comments

nzpr committed:
Move DagRepresentation pure calls on a class

bors[bot] committed:
Merge #3662

3662: Refactor BlockAPI r=stanislavlyalin a=stanislavlyalin



Co-authored-by: stanislavlyalin <lyalinstas@gmail.com>

tgrospic committed:
Removed legacy config for RSpace (history and cold store)

tgrospic committed:
Removed unused block errors (Processed, CasperIsBusy, DeployNotSigned)

tgrospic committed:
Removed block parents check and errors

tgrospic committed:
Remove follow justification check and errors

tgrospic committed:
Removed neglected equivocation block error

tgrospic committed:
Removed unused admissible and ignorable equivocation block errors

nzpr committed:
Rebase to latest dev

bors[bot] committed:
Merge #3666

3666: Prepare pure state for BlockDagRepresentation r=nzpr a=nzpr



Co-authored-by: nutzipper <1746367+nzpr@users.noreply.github.com>

tgrospic committed:
Remove unused functions (isInDag, computeCodeHash, getRholangDeployParams)

tgrospic committed:
Test SingleParentCasper cleanup

tgrospic committed:
Remove invalid block hash and shard ID from slashing

tgrospic committed:
Remove Casper from BlockProcessor check if block is of interest

tgrospic committed:
Removed casperSnapshot from Casper

tgrospic committed:
Removed approved block from Casper

tgrospic committed:
Remove version from Casper

tgrospic committed:
Removed block contains methods from Casper

tgrospic committed:
Remove EngineCell dependency from updateForkChoiceTipsIfStuck

tgrospic committed:
Remove validate method from Casper

tgrospic committed:
Remove handle valid/invalid blocks from Casper and lastFinalizedBlock

tgrospic committed:
Remove dag representation, fetch dependencies and getDependencyFreeFromBuffer from Casper

tgrospic committed:
Remove Casper from block processing input stream

tgrospic committed:
Simplify implicits in new BlockAPI implementation

tgrospic committed:
Swithing to new BlockAPI implementation, removing old BlockAPI object with functions

tgrospic committed:
Rename BlockAPI_v2 to BlockApi (BlockReportApi)

tgrospic committed:
Fix deploy phlo limit test

tgrospic committed:
Fix propose trigger function supplied to block API

tgrospic committed:
Remove unused parent shard from shard config

tgrospic committed:
Fix auto propose after deploy

tgrospic committed:
Remove unnecessary dependencies from deploy gRPC builder

tgrospic committed:
Temporary disable integration tests for slashing (invalid block and justification)

tgrospic committed:
Remove Casper trait and implementations completely

tgrospic committed:
Rename Casper and MultiParentCasperImpl (CasperSnapshot, MultiParentCasper)

tgrospic committed:
Simplify deploy ParsingError (delete unused variations)

tgrospic committed:
Fix succeed at slashing test (deploy shard ID)

tgrospic committed:
Fix for genesis ceremony test (wait for approve block)

tgrospic committed:
Temp fix for deploy parsing error message

bors[bot] committed:
Merge #3672

3672: Block processing refactoring (switch to new BlockApi, Casper cleanup) r=stanislavlyalin a=tgrospic



Co-authored-by: Tomislav Grospic <grospic@gmail.com>

tgrospic committed:
Remove unused fairDispatcher

tgrospic committed:
Move CasperPacketHandler to protocol folder

tgrospic committed:
Move gRPC client (deploy/propose service) to protocol/client folder

tgrospic committed:
Remove unused DagOperations object

tgrospic committed:
Move CommUtil to protocol folder

tgrospic committed:
Move Runtime syntax to syntax sub-folder

tgrospic committed:
Move rholang files from util to rholang folder

tgrospic committed:
Move rholang tests to rholang folder (from util), update CI workflow

tgrospic committed:
Remove unused Clique object and tests

tgrospic committed:
Fix import in Main

bors[bot] committed:
Merge #3678

3678: Casper rholang cleanup r=tgrospic a=tgrospic



Co-authored-by: Tomislav Grospic <grospic@gmail.com>

tgrospic committed:
Remove unused dummy PacketHandler

tgrospic committed:
Refactor of genesis ceremony with network packer receiver, removed Engine/EngineCell

tgrospic committed:
Change LFS chunk size to 750 nodes

tgrospic committed:
Rename transport stream identifier StreamMsgId

tgrospic committed:
Move transport layer stream1 method as extension

tgrospic committed:
Fix network message receiver

tgrospic committed:
Remove old genesis ceremony config parameters and deploy timestamp

tgrospic committed:
Remove config for round robin dispatcher

tgrospic committed:
Remove unused supply parameter from REV generator for genesis and unused timestamp for blessed deploys

tgrospic committed:
Fix removed CLI options in integration tests

tgrospic committed:
Remove unused functions from ProtoUtil and Validate block

tgrospic committed:
Move create genesis function to CasperLaunch, remove ApprovedBlock completely

tgrospic committed:
Remove waiting on approved block in genesis integration test

tgrospic committed:
Refactor node startup, sync and running mode

tgrospic committed:
Rename CasperLaunch to NodeLaunch

tgrospic committed:
Move reporting file into reporting sub-folder

bors[bot] committed:
Merge #3679

3679: Genesis ceremony cleanup r=tgrospic a=tgrospic



Co-authored-by: Tomislav Grospic <grospic@gmail.com>
